### PR TITLE
Subject import API

### PR DESCRIPTION
**Pull request by marten**
Describe your change here.

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

File: app/models/subject_set_import/processor.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+class SubjectSetImport::Processor
+  attr_reader :subject_set, :uploader
+
+  def initialize(subject_set, uploader)
+    @subject_set = subject_set
+    @uploader = uploader
+  end
+
+  def import(uuid, attributes)
+    subject = subject_set.subjects.where(external_id: uuid).first_or_initialize
```
[the remark this answers is not in the document]
Comment (camallen) [reply app/models/subject_set_import/processor.rb]: How is the UUID here ensuring unique subject lookups? Or will it clobber subject information if it finds one with an existing `external_id`?

File: app/serializers/subject_set_import_serializer.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class SubjectSetImportSerializer
+  include RestPack::Serializer
+  include CachedSerializer
+
+  attributes :id, :href, :created_at, :updated_at, :source_url
+  can_include :subject_set, :user
+
+  can_filter_by :subject_set, :user
```
Comment (camallen) [inline app/serializers/subject_set_import_serializer.rb]: do these have FK indexes for filtering?

File: spec/models/subject_set_import_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  describe 'scope_for' do
+    it 'allows project owners to create an import' do
+      api_user = ApiUser.new(user)
+      import = described_class.new
+      import.subject_set = SubjectSet.link_to_resource(import, api_user).where(id: subject_set.id).first
+      import.user = User.link_to_resource(import, api_user).where(id: user.id).first
+      import.save!
+    end
+
+    it 'disallows peasants to create an import' do
```
Comment (camallen) [inline spec/models/subject_set_import_spec.rb]: pheasants maybe?

File: app/models/subject_set_import/processor.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (marten) [reply app/models/subject_set_import/processor.rb]: Yeah, it'll overwrite. UUID is meant so that an external party doesn't have to keep track of what they've already got uploaded. It's their responsibility to make sure that it's generated consistently. Thinking about it, maybe we can also support CSVs without a UUID column, in which case we just create new subjects for every row.

I should also rename this. In my own Python example notebook I ended up calculating the SHA1 of the images and using that as the "UUID". Panoptes doesn't really care if it's a UUID or some other identifier (hence the `external_id`), so why should the CSV file?

File: app/controllers/api/v1/subject_set_imports_controller.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class Api::V1::SubjectSetImportsController < Api::ApiController
+  require_authentication :create, scopes: [:project]
+
+  resource_actions :index, :show, :create
```
[the remark this answers is not in the document]
Comment (marten) [reply app/controllers/api/v1/subject_set_imports_controller.rb]: How so? I searched and couldn't see any other `require_authentication` lines referencing `index` or `show`, but it's likely I'm misunderstanding you or the code.

File: app/schemas/subject_update_schema.rb [hunk 5]
```diff
@@ -4,6 +4,10 @@ class SubjectUpdateSchema < JsonSchema
     description "A peice of media to classifiy"
     additional_properties false
 
+    property "external_id" do
```
[the remark this answers is not in the document]
Comment (marten) [reply app/schemas/subject_update_schema.rb]: No, I guess not.

File: app/serializers/subject_set_import_serializer.rb (hunk 2 above)
Comment (marten) [reply app/serializers/subject_set_import_serializer.rb]: SubjectSet did, I've added user

File: app/controllers/api/v1/subject_set_imports_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (camallen) [reply app/controllers/api/v1/subject_set_imports_controller.rb]: If it's a protected resource that you only want to show to logged in users then you can add the `require_authentication` checks, i assume this is. That will then run before any authorization code which should only return what you are allowed to access,`404` or `[]`. It adds another layer of access checks that will be linked to the token scopes as well (i.e. i am authenticated and have a token but don't have the correct scopes to access that resource, limited rights oauth application, etc). It's up to you how if you want to report unauthenticated or rely on authorization scopes for non-logged, for this resource.

See https://github.com/zooniverse/Panoptes/blob/af2e49ea802c0636ef7902605893f7bd2cbf1b02/app/controllers/api/v1/classifications_controller.rb#L5 and https://github.com/zooniverse/Panoptes/blob/af2e49ea802c0636ef7902605893f7bd2cbf1b02/app/controllers/api/v1/memberships_controller.rb#L3 for examples of this authentication check on controllers.

[the remark this answers is not in the document]
Comment (marten) [reply app/controllers/api/v1/subject_set_imports_controller.rb]: So how do I ensure the tests check that none of this is publicly visible, or visible to other project owners? Can't see anything to that effect in the tests for Membership or Classification

[the remark this answers is not in the document]
Comment (camallen) [reply app/controllers/api/v1/subject_set_imports_controller.rb]: Not explicitly via the look, i thought this was tested in the shared controller helper modules, e.g. indexable, showable, etc though it appears to only be tested in create / destroy actions. 

The authentication and scopes are tested in the controllers as they will generally fail if authentication is required and not setup in the specs, that said it's inverted as it looks that the tests can pass if a `require_authentication` is never used on a controller.

File: app/schemas/subject_create_schema.rb [hunk 6]
```diff
@@ -5,6 +5,10 @@ class SubjectCreateSchema < JsonSchema
     required "links", "locations"
     additional_properties false
 
+    property "external_id" do
```
[the remark this answers is not in the document]
Comment (marten) [reply app/schemas/subject_create_schema.rb]: I figured it's just a nice thing to do to expose it, since it's also serialized back? But don't care too much, no thought was really going into this on my part.

**Comment by marten** [comment]
@camallen okay, ready for some more reviewing

File: app/schemas/subject_create_schema.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (marten) [reply app/schemas/subject_create_schema.rb]: Hmm, well there's no check to make sure that `external_id` remains unique within a subject set. I guess with this added to the normal API you could make it so that on a subsequent Import it's no longer specified which subject it'll upsert. Then again, using `update_links` you could add existing subjects to a set and create clashes as well. 

Do you think it's worth it to add stuff to enforce uniqueness (either in Rails or PG)? It'll be a non-trivial query, and we don't really expect people that use Imports to also do other uploading into that same set. 

Alternatively, we could mark the subject set as "using_imports" and disallow normal API linking?

[the remark this answers is not in the document]
Comment (camallen) [reply app/schemas/subject_create_schema.rb]: I don't think we will need it, let's just be aware of this if issues arise.
